### PR TITLE
[ Kernel ] FP8 Dynamic-Per-Token Quant Kernel

### PR DESCRIPTION
**Pull request by varun-sundar-rabindranath**
Add support for dynamic-per-token fp8 quantization.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/_custom_ops.py [hunk 1]
```diff
@@ -320,6 +320,17 @@ def scaled_fp8_quant(
     return output, scale
 
 
+def dynamic_per_token_scaled_fp8_quant(
+        input: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (varun-sundar-rabindranath) [inline vllm/_custom_ops.py:324]: To fit the existing interface this should probably be merged up with the `scaled_fp8_quant`. How do we want to handle this ? 
I believe the `dynamic-per-token` vs `dynamic-per-tensor` model quantization scheme will come from the model itself and we can plumb it down into the `scaled_fp8_quant` function. I'd appreciate any references into will how to plumb this down. Thanks. 
@robertgshaw2-neuralmagic @comaniac @tlrmchlsmth  ?

File: tests/kernels/test_int8_quant.py [hunk 2]
```diff
@@ -3,6 +3,8 @@
```
Comment (varun-sundar-rabindranath) [inline tests/kernels/test_int8_quant.py:3]: Mostly refactors in this file!

File: csrc/quantization/fp8/common.cu [hunk 3]
```diff
@@ -118,12 +115,62 @@ __global__ void scaled_fp8_quant_kernel(c10::Float8_e4m3fn* __restrict__ out,
...
+    c10::Float8_e4m3fn* __restrict__ out, float* __restrict__ scale,
+    scalar_t const* __restrict__ input, const int hidden_size) {
+  int const tid = threadIdx.x;
+  int const token_idx = blockIdx.x;
+  float absmax_val = 0.0f;
+
+  for (int i = tid; i < hidden_size; i += blockDim.x) {
+    float const x = static_cast<float>(input[token_idx * hidden_size + i]);
+    absmax_val = max(absmax_val, fabs(x));
+  }
```
Comment (mgoin) [inline csrc/quantization/fp8/common.cu:149]: Could this be vectorized or unrolled?

File: vllm/_custom_ops.py (hunk 1 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/_custom_ops.py:324]: I think this looks good as is. I can handle the plumbing from here in a follow up with some end-to-end model tests

File: csrc/quantization/fp8/common.cu [hunk 4]
```diff
@@ -118,12 +115,62 @@ __global__ void scaled_fp8_quant_kernel(c10::Float8_e4m3fn* __restrict__ out,
...
+  float const block_absmax_val_maybe = blockReduceMax(absmax_val);
+  __shared__ float block_absmax_val;
+  if (tid == 0) {
+    block_absmax_val = block_absmax_val_maybe;
+    scale[token_idx] = block_absmax_val / FP8_E4M3_MAX;
+  }
+  __syncthreads();
+
+  float const inverted_scale = FP8_E4M3_MAX / block_absmax_val;
+  bool const vectorize_conversions = hidden_size % 4 == 0;
```
Comment (tlrmchlsmth) [inline csrc/quantization/fp8/common.cu:160]: Why only vectorize conversions if the hidden_size is divisible by 4? `scaled_fp8_conversion_vec` has a loop to "Handle the remaining elements if num_elems is not divisible by 4"

File: tests/kernels/quant_utils.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+def ref_dynamic_per_token_quant(x: torch.tensor,
+                                quant_dtype: torch.dtype) \
+        -> Tuple[torch.tensor, torch.tensor]:
+
+    assert quant_dtype in [torch.int8, torch.float8_e4m3fn]
+    qtype_traits = torch.iinfo(quant_dtype) if quant_dtype == torch.int8 \
+            else torch.finfo(quant_dtype)
+    qtype_max = as_float32_tensor(qtype_traits.max)
+
+    # For fp8, inorder to match the cuda kernel output, we have to do exactly
```
Comment (tlrmchlsmth) [inline tests/kernels/quant_utils.py:18]: nit: inorder -> in order

File: tests/kernels/quant_utils.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+# ref_dynamic_per_token_quant, when we have a dynamic_per_tensor int8 quant
+# kernel
+def ref_dynamic_per_tensor_fp8_quant(x: torch.tensor) \
+                    -> Tuple[torch.tensor, torch.tensor]:
+
+    fp8_traits = torch.finfo(torch.float8_e4m3fn)
+    fp8_max = as_float32_tensor(fp8_traits.max)
+    one = as_float32_tensor(1.0)
+
+    # For fp8, inorder to match the cuda kernel output, we have to do exactly
```
Comment (tlrmchlsmth) [inline tests/kernels/quant_utils.py:47]: ditto: inorder -> in order

**Review by tlrmchlsmth — COMMENTED** [review]
I left a few in-line comments. Mostly nits, but I also had a one question on the vectorization when the hidden size isn't divisible by 4.

File: csrc/quantization/fp8/common.cu (hunk 4 above)
Comment (varun-sundar-rabindranath) [reply csrc/quantization/fp8/common.cu:160]: The input and output pointers need to be aligned  8-byte and 4-byte addresses respectively.  Each row of the input is processed by a single block. 
The FP8 output data-type requires that hidden-size (each token/row of the output has hidden-size elements) be a multiple of 4, otherwise the 2nd, 3rd and 4th output rows will no the 4-byte aligned. The argument is similar for float16, bfloat16, float32 input datatypes.

File: vllm/_custom_ops.py (hunk 1 above)
Comment (comaniac) [reply vllm/_custom_ops.py:324]: Definitely prefer to have a unified API.

**Review by comaniac — APPROVED** [review]
Approved to unblock.
1. Same question @tlrmchlsmth that why not use vectorization for non-dividable problem size?
2. Better to unify the API. Ok to move to the next PR.

File: csrc/ops.h [hunk 7]
```diff
@@ -129,6 +129,10 @@ void static_scaled_fp8_quant(torch::Tensor& out, torch::Tensor& input,
 void dynamic_scaled_fp8_quant(torch::Tensor& out, torch::Tensor& input,
                               torch::Tensor& scale);
 
+void dynamic_per_token_scaled_fp8_quant(torch::Tensor& out,
+                                        torch::Tensor& input,
+                                        torch::Tensor& scale);
```
[the remark this answers is not in the document]
Comment (varun-sundar-rabindranath) [reply csrc/ops.h]: Done.

File: csrc/quantization/fp8/common.cu (hunk 3 above)
Comment (varun-sundar-rabindranath) [reply csrc/quantization/fp8/common.cu]: Yes. I overlooked it. Fixed it now 👍  Thanks.

File: csrc/quantization/fp8/common.cu (hunk 4 above)
Comment (varun-sundar-rabindranath) [reply csrc/quantization/fp8/common.cu]: @comaniac

Comment (tlrmchlsmth) [reply csrc/quantization/fp8/common.cu]: Why didn't we run into this with the original `scaled_fp8_quant_kernel`?

Comment (varun-sundar-rabindranath) [reply csrc/quantization/fp8/common.cu]: the `scaled_fp8_quant_kernel` has a single scale that is applied to all of the input values and it doesn't have to process row-by-row. For quant purposes, that kernel can treat the inputs and output as one big row of values (the "first row" case from above).

**Review by tlrmchlsmth — APPROVED** [review]
Thanks!
